### ISSUE TITLE
Flate decompression error: Error -5 while decompressing data: incomplete or truncated stream

### ISSUE THREAD

**Issue by wilbowma** [pts/pdfsizeopt#120]
Using most recent `pdfsizeopt` on macOS on the file located here: https://github.com/pts/pdfsizeopt/files/10817079/okasaki1996.-.Purely.Functional.Data.Structures.pdf

```
info: This is pdfsizeopt ZIP rUNKNOWN size=69734.
info: prepending to PATH: /usr/local/Cellar/pdfsizeopt/358d2b/libexec
info: loading PDF from: okasaki1996 - Purely Functional Data Structures.pdf
info: loaded PDF of 1324780 bytes
info: separated to 1092 objs + xref + trailer
info: parsed 1092 objs
info: eliminated 20 unused objs, depth=11
info: found 0 Type1 fonts loaded
info: found 2 Type1C fonts loaded
info: writing Type1CParser (3806 font bytes) to: /var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.parse.tmp.ps
info: using Ghostscript TMPDIR=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T TEMP=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T /usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt_gs/gs: GPL Ghostscript 9.05 (2012-02-08)
info: executing Type1CParser with Ghostscript: TMPDIR=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T TEMP=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T /usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt_gs/gs -q -P- -dNOPAUSE -dBATCH -sDEVICE=nullpage -sDataFile=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.parsedata.tmp.ps -f /var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.parse.tmp.ps
Type1CParser: using interpreter GPL Ghostscript 905 20120208
Type1CParser: all OK
info: parsed 2 Type1C fonts
info: writing Type1CGenerator (9196 bytes in 2 fonts) to: /var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.gen.tmp.ps
info: executing Type1CGenerator with Ghostscript: TMPDIR=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T TEMP=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T /usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt_gs/gs -q -P- -dNOPAUSE -dBATCH -sDEVICE=pdfwrite -dPDFSETTINGS=/printer -dColorConversionStrategy=/LeaveColorUnchanged -sOutputFile=/var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.gen.tmp.pdf -f /var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.gen.tmp.ps
Type1CGenerator: using interpreter GPL Ghostscript 905 20120208
Type1CGenerator: all OK
info: loading PDF from: /var/folders/4j/rhfkfpd17xddfbbp1f6twmnr0000gn/T/psotmp.81606.conv.gen.tmp.pdf
info: loaded PDF of 7743 bytes
info: separated to 18 objs + xref + trailer
info: parsed 18 objs
info: found 2 fonts loaded
info: optimized Type1C fonts to form 4216 bytes to 4174 bytes (99%)
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/__main__.py", line 1, in <module>
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/m.py", line 6, in <module>
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/pdfsizeopt/main.py", line 5629, in main
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/pdfsizeopt/main.py", line 4005, in ConvertInlineImagesToXObjects
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/pdfsizeopt/main.py", line 1782, in DetectInlineImage
  File "/usr/local/Cellar/pdfsizeopt/358d2b/libexec/pdfsizeopt/pdfsizeopt/main.py", line 1986, in GetUncompressedStream
pdfsizeopt.main.FilterError: Flate decompression error: Error -5 while decompressing data: incomplete or truncated stream
```

**Comment by pts** [comment]
Thank you for reporting this! Until the issue is fixed, you can use pdfsizeopt with `--do-optimize-images=no` as a workaround.

**Comment by pts** [comment]
Could you please upload the input PDF file again?

**Comment by pts** [comment]
Fixed in 577951961957f0c2deec99e55bccd9d6bdf2bb76. The problem was caused by the *PermissiveZlibDecompress* function in pdfsizeopt rejecting a zlib stream ending with Z_SYNC_FLUSH (and no Adler-32 checksum) rather than Z_FINISH.
